starlark/int.go

```diff
@@ -142,75 +138,161 @@ var (
 	maxint64 = new(big.Int).SetInt64(math.MaxInt64)
 )
 
-func (i Int) String() string { return i.bigint.String() }
+func (i Int) Text(base int) string {
+	if i.bigint != nil {
+		return i.bigint.Text(base)
+	}
+	return strconv.FormatInt(i.val, base)
+}
+func (i Int) String() string { return i.Text(10) }
 func (i Int) Type() string   { return "int" }
 func (i Int) Freeze()        {} // immutable
 func (i Int) Truth() Bool    { return i.Sign() != 0 }
 func (i Int) Hash() (uint32, error) {
 	var lo big.Word
-	if i.bigint.Sign() != 0 {
+	if i.bigint != nil && i.bigint.Sign() != 0 {
 		lo = i.bigint.Bits()[0]
+	} else {
+		lo = big.Word(i.val)
 	}
 	return 12582917 * uint32(lo+3), nil
 }
-func (x Int) CompareSameType(op syntax.Token, y Value, depth int) (bool, error) {
-	return threeway(op, x.bigint.Cmp(y.(Int).bigint)), nil
+func (x Int) CompareSameType(op syntax.Token, v Value, depth int) (bool, error) {
+	y := v.(Int)
+	if x.bigint != nil && y.bigint != nil {
+		return threeway(op, x.big().Cmp(y.big())), nil
+	}
+	return threeway(op, int(x.val-y.val)), nil
 }
 
 // Float returns the float value nearest i.
 func (i Int) Float() Float {
-	// TODO(adonovan): opt: handle common values without allocation.
-	f, _ := new(big.Float).SetInt(i.bigint).Float64()
-	return Float(f)
-}
-
-func (x Int) Sign() int      { return x.bigint.Sign() }
-func (x Int) Add(y Int) Int  { return Int{new(big.Int).Add(x.bigint, y.bigint)} }
-func (x Int) Sub(y Int) Int  { return Int{new(big.Int).Sub(x.bigint, y.bigint)} }
-func (x Int) Mul(y Int) Int  { return Int{new(big.Int).Mul(x.bigint, y.bigint)} }
-func (x Int) Or(y Int) Int   { return Int{new(big.Int).Or(x.bigint, y.bigint)} }
-func (x Int) And(y Int) Int  { return Int{new(big.Int).And(x.bigint, y.bigint)} }
-func (x Int) Xor(y Int) Int  { return Int{new(big.Int).Xor(x.bigint, y.bigint)} }
-func (x Int) Not() Int       { return Int{new(big.Int).Not(x.bigint)} }
-func (x Int) Lsh(y uint) Int { return Int{new(big.Int).Lsh(x.bigint, y)} }
-func (x Int) Rsh(y uint) Int { return Int{new(big.Int).Rsh(x.bigint, y)} }
+	if i.bigint != nil {
+		// TODO(adonovan): opt: handle common values without allocation.
+		f, _ := new(big.Float).SetInt(i.bigint).Float64()
+		return Float(f)
+	}
+	return Float(float64(i.val))
+}
+
+func (x Int) Sign() int {
+	switch {
+	case x.bigint != nil:
+		return x.bigint.Sign()
+	case x.val > 0:
```

Comment:
This version is less branchy:

if x.bigint != nil {
   return x.bigint.Sign()
}
return signum64(x.val)

(See existing signum function.)